File: wechat-miniprogram/cloudtemplates/bridge-api/bridge_trainer/cards.py
from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Iterable

SUITS = ["S", "H", "D", "C"]
SUIT_SYMBOLS = {"S": "♠", "H": "♥", "D": "♦", "C": "♣"}
RANKS = ["A", "K", "Q", "J", "10", "9", "8", "7", "6", "5", "4", "3", "2"]
RANK_ORDER = {rank: index for index, rank in enumerate(RANKS)}
SUIT_NAMES = {"S": "黑桃", "H": "红心", "D": "方块", "C": "梅花"}


@dataclass(frozen=True, order=True)
class Card:
    suit: str
    rank: str

    def label(self) -> str:
        return f"{SUIT_SYMBOLS[self.suit]}{self.rank}"


Hand = list[Card]


def new_deck() -> list[Card]:
    return [Card(suit, rank) for suit in SUITS for rank in RANKS]


def deal(seed: int | None = None) -> dict[str, Hand]:
    deck = new_deck()
    rng = random.Random(seed)
    rng.shuffle(deck)
    players = ["N", "E", "S", "W"]
    return {player: sort_hand(deck[i * 13 : (i + 1) * 13]) for i, player in enumerate(players)}


def sort_hand(hand: Iterable[Card]) -> Hand:
    return sorted(hand, key=lambda card: (SUITS.index(card.suit), RANK_ORDER[card.rank]))


def cards_by_suit(hand: Iterable[Card]) -> dict[str, list[Card]]:
    grouped = {suit: [] for suit in SUITS}
    for card in sort_hand(hand):
        grouped[card.suit].append(card)
    return grouped


def format_hand_lines(hand: Iterable[Card]) -> list[str]:
    grouped = cards_by_suit(hand)
    lines: list[str] = []
    for suit in SUITS:
        ranks = " ".join(card.rank for card in grouped[suit]) or "—"
        lines.append(f"{SUIT_SYMBOLS[suit]} {ranks}")
    return lines

File: wechat-miniprogram/utils/bridge_trainer/cards.js
"use strict";

const { PythonRandom } = require("./random");

const SUITS = ["S", "H", "D", "C"];
const SUIT_SYMBOLS = { S: "♠", H: "♥", D: "♦", C: "♣" };
const RANKS = ["A", "K", "Q", "J", "10", "9", "8", "7", "6", "5", "4", "3", "2"];
const RANK_ORDER = {};
RANKS.forEach((rank, index) => {
  RANK_ORDER[rank] = index;
});
const SUIT_NAMES = { S: "黑桃", H: "红心", D: "方块", C: "梅花" };

function createCard(suit, rank) {
  return {
    suit,
    rank,
    label() {
      return `${SUIT_SYMBOLS[this.suit]}${this.rank}`;
    },
  };
}

function new_deck() {
  const deck = [];
  for (const suit of SUITS) {
    for (const rank of RANKS) {
      deck.push(createCard(suit, rank));
    }
  }
  return deck;
}

function sort_hand(hand) {
  return hand.slice().sort(function (a, b) {
    const suitDiff = SUITS.indexOf(a.suit) - SUITS.indexOf(b.suit);
    if (suitDiff !== 0) {
      return suitDiff;
    }
    return RANK_ORDER[a.rank] - RANK_ORDER[b.rank];
  });
}

function deal(seed) {
  const deck = new_deck();
  const rng = new PythonRandom(seed === undefined || seed === null ? Date.now() : seed);
  rng.shuffle(deck);
  const players = ["N", "E", "S", "W"];
  const hands = {};
  for (let i = 0; i < players.length; i += 1) {
    hands[players[i]] = sort_hand(deck.slice(i * 13, (i + 1) * 13));
  }
  return hands;
}

function cards_by_suit(hand) {
  const grouped = { S: [], H: [], D: [], C: [] };
  for (const card of sort_hand(hand)) {
    grouped[card.suit].push(card);
  }
  return grouped;
}

function format_hand_lines(hand) {
  const grouped = cards_by_suit(hand);
  const lines = [];
  for (const suit of SUITS) {
    const ranks = grouped[suit].map((card) => card.rank).join(" ") || "—";
    lines.push(`${SUIT_SYMBOLS[suit]} ${ranks}`);
  }
  return lines;
}

module.exports = {
  SUITS,
  SUIT_SYMBOLS,
  RANKS,
  RANK_ORDER,
  SUIT_NAMES,
  createCard,
  new_deck,
  deal,
  sort_hand,
  cards_by_suit,
  format_hand_lines,
};
